website/payments/api/v2/admin/views.py

```diff
@@ -0,0 +1,130 @@
+from django.apps import apps
+from django.http import Http404
+from django.utils.translation import gettext_lazy as _
+from oauth2_provider.contrib.rest_framework import IsAuthenticatedOrTokenHasScope
+from rest_framework import status, serializers
+from rest_framework.exceptions import PermissionDenied, ValidationError
+from rest_framework.generics import get_object_or_404
+from rest_framework.response import Response
+from rest_framework.reverse import reverse
+from rest_framework.settings import api_settings
+
+from payments import services, payables, NotRegistered
+from payments.api.v2 import filters
+from payments.api.v2.admin.serializers.payable_create import PayableCreateSerializer
+from payments.api.v2.admin.serializers.payable_detail import PayableSerializer
+from payments.api.v2.serializers import PaymentSerializer
+from payments.exceptions import PaymentError
+from payments.models import Payment, PaymentUser
+from thaliawebsite.api.v2.admin import (
+    AdminListAPIView,
+    AdminCreateAPIView,
+    AdminRetrieveAPIView,
+    AdminDestroyAPIView,
+    AdminUpdateAPIView,
+)
+
+
+class PaymentListCreateView(AdminListAPIView):
+    queryset = Payment.objects.all()
+    serializer_class = PaymentSerializer
+
+    required_scopes = ["payments:admin"]
+    filter_backends = (
+        framework_filters.OrderingFilter,
+        filters.CreatedAtFilter,
+        filters.PaymentTypeFilter,
+    )
+    ordering_fields = ("created_at",)
+
+
+class PaymentDetailView(AdminRetrieveAPIView, AdminDestroyAPIView):
+    queryset = Payment.objects.all()
+    serializer_class = PaymentSerializer
+    permission_classes = [IsAuthenticatedOrTokenHasScope]
+    required_scopes = ["payments:admin"]
+
+
+class PayableBaseView:
+    required_scopes = ["payments:admin"]
+
+    def get_permissions(self):
+        return [IsAuthenticatedOrTokenHasScope()]
+
+    def get_payable(self):
+        app_label = self.kwargs["app_label"]
+        model_name = self.kwargs["model_name"]
+        payable_pk = self.kwargs["payable_pk"]
+
+        try:
+            payable_model = apps.get_model(app_label=app_label, model_name=model_name)
+            payable = payables.get_payable(
+                get_object_or_404(payable_model, pk=payable_pk)
+            )
+        except (LookupError, NotRegistered) as e:
+            raise serializers.ValidationError(
+                {api_settings.NON_FIELD_ERRORS_KEY: [_("Payable model not found")]}
+            ) from e
+
+        if not payable.can_manage_payment(self.request.member):
+            raise PermissionDenied(
+                detail=_("You do not have permission to perform this action.")
+            )
+
+        return payable
+
+
+class PayableDetailView(
+    PayableBaseView, AdminRetrieveAPIView, AdminUpdateAPIView, AdminDestroyAPIView
+):
+    """View that allows you to manipulate the payment for the payable. Permissions of this view are based on the payable."""
```

Comment:
I don't think I fully understand this. Creating a payment happens by POSTing to a payable? I myself thought we would POST an app name, model name and pk to a general endpoint to create new payments. But this is fine as well I think

Reply:
I needed to find a way to use the payables, and also allow for actually CRUD operations for the payments. Just sending the information to /api/v2/payments does not allow you to create a payment manually anymore.

This is a payment that will be created based on the payable, and we need a way to remove them based on the payable as well. Django rest framework does not really support DELETE requests with a body. So I opted to add the information to the URL and actually make this a REST entity. It is now a PUT (thus update) to the payable that will create and add the payment.

Reply:
There is no need for DELETE's with a body right? 

Anyway, I do see your idea. It's just different from what I had in mind myself